### PR TITLE
improving queue trigger handling of invalid messages

### PR DESCRIPTION
**Pull request by brettsam**
Addresses #768

### CODE REVIEW

**Comment by mathewc** [comment]
I don't think we want to do the change in QueueProcessor. We don't want to automatically move these messages to poison queue. E.g. We want to support binding to CloudQueueMessage rather than string - in this case that should work. I.e. if the user isn't trying to bind as a string, we shouldn't be enforcing that its a string, or doing an AsString.

I think we need to simply make our checks in [StorageQueueMessageExtensions](https://github.com/Azure/azure-webjobs-sdk/blob/master/src/Microsoft.Azure.WebJobs.Host/Queues/StorageQueueMessageExtensions.cs#L25) more robust. We should handle the FormatException there as well. That will let the message go through, and it will get normal poison queue handling.

**Comment by brettsam** [comment]
Good point about letting CloudQueueMessage get through; I hadn't tested that. I think you're right. Handling it there seems to throw an exception within the function invocation (likely in a converter), which then triggers the proper error and poison-queue flow. Let me play with it a bit more and I'll submit a revised PR.

**Comment by mathewc** [comment]
I do like your idea of having a way of identifying a message like this that will **never** succeed - in such cases, doing retries will never work. However I think that is a potential improvement that would require more thought on the right way to do it across the board (e.g. blob trigger has poison queue handling as well). So lets keep this one scoped.

**Comment by brettsam** [comment]
The code change is simpler now. I ended up writing an e2e test since I wanted to ensure that this worked through the whole pipeline -- that CloudQueueMessage params got the message and that string params threw and eventually went to the poison queue.

File: src/Microsoft.Azure.WebJobs.Host/Queues/StorageQueueMessageExtensions.cs [hunk 1]
```diff
@@ -27,6 +27,10 @@ public static string TryGetAsString(this IStorageQueueMessage message)
                 // when the message type is binary, AsString will throw if the bytes aren't valid UTF-8.
                 value = null;
             }
+            catch (FormatException)
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Queues/StorageQueueMessageExtensions.cs]: nit: 

```catch (Exception ex)
{
    if (ex is DecoderFallbackException || ex is FormatException)
    {
        value = null;
        return;
    }
    else
    {
        throw;
    }
}
```

**Comment by mathewc** [comment]
:ship:

File: src/Microsoft.Azure.WebJobs.Host/Queues/StorageQueueMessageExtensions.cs (hunk 1 above)
Comment (fabiocav) [inline src/Microsoft.Azure.WebJobs.Host/Queues/StorageQueueMessageExtensions.cs]: Exception filters clean the proposed code, but we'd need to be able to use C# 6 in the project :)

Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Queues/StorageQueueMessageExtensions.cs]: Yeah, we can't do that yet until we move the SDK and Dashboard projects from TeamCity to AppVeyor. If somebody magically made that happen while I'm away on vacation, I'd buy them a case of beer :)

**Pull request merged**